### PR TITLE
Structured EEPROM save & restore for all keyboards

### LINKED ISSUE #795: Chrysalis 0.9.0 wrote the wrong keyboard layout on Keyboardio Atreus during firmware update
**Describe the bug**
Updating firmware on my _Keyboardio Atreus_ completely messed up my keyboard layout - the columns were shifted, i.e. _Q_ key produced _W_ when pressed, _C_ produced _V_, etc. It also obliterated all of my custom configuration. Is it possible that it detected and/or used the layout for _Model 01_?

**To Reproduce**
1. Plug in _Keyboardio Atreus_.
2. Open _Chrysalis_.
3. Setect a keyboard.
4. Connect.
5. Firmware Update.
6. Selected firmware - Chrysalis 0.9.0 default.
7. Press bottom left key.
8. Update.

**Expected behavior**
Firmware updates **without** touching the keyboard layout, including local modifications, unless _Restore to factory defaults when flashing_ explicitly selected.

**Desktop:**
 - OS: macOS 10.15.7
 - Chrysalis Version: 0.9.0

**Additional context**
Yes, I'm aware that version _0.9.1_ has been released shortly afterwards but that's what was, and (so far) still is, available via _homebrew_ - also, looking at the commits between the two versions, I see that the changes are cosmetic, i.e. _JS_ and _CSS_.

I have another Atreus which I can update the firmware on using this or another machine (macOS 12.4) with the same or newer version of _Chrysalis_ if need be.

BTW, I had to reset EEPROM to factory defaults and redo my custom layout by hand - I exported it to a file now (attached). Debug bundle also attached.

[rjc-atreus_custom_layout.json.txt](https://github.com/keyboardio/Chrysalis/files/8719372/rjc-atreus_custom_layout.json.txt)
[rjc-chrysalis-debug.bundle.json.txt](https://github.com/keyboardio/Chrysalis/files/8719373/rjc-chrysalis-debug.bundle.json.txt)

### PR DESCRIPTION
**Pull request by algernon**
A major problem - as highlighted in #795 - with our flashing process is that during the firmware update, if the EEPROM layout changes, bad things will happen. In the past we tried our best to not introduce layout changes in the firmware we ship with Chrysalis, but one such change slipped through.

On one hand, we could back that out, and re-add the feature without a layout change, but a whole lot of people already flashed those firmware. We'd break the eeprom layout for some people no matter what we do.

So instead of that, we chose to fix the problem at its root: instead of not touching EEPROM at all (on AVR) or restoring it as-is (on GD32), with this pull request, we'll save & restore it in a structured fashion: we save all the settings the built-in plugins can have (even for plugins not directly supported by Chrysalis), and restore them after flashing.

We do this for all keyboards, even those that do not clobber EEPROM on flashing. This way, we can change the EEPROM layout in the firmware we ship with, and things will still work after, because we migrate to the new layout with the post-flash restore.

Fixes #795.

### CODE REVIEW

**Comment by algernon** [comment]
There's an edge case here that's worth mentioning: when reconnecting after a flash, we simply wait until we find a keyboard that matches the VID/PID of the one we flashed. If there are multiple keyboards of the same type plugged in, there's a good chance we'll connect and restore (& clobber) the one we *didn't* flash.

The reconnect logic doesn't require that the keyboard is something that just came back, it just waits for any suitable keyboard to appear, or be present to begin with.

As such, it is **not** safe to flash with multiple keyboards of the same kind being plugged in.

Perhaps the FirmwareUpdate screen should have a look at connected keyboards, and if it detects a possible clash, refuse to flash? The other option is to make sure that the keyboard we reconnect to is one that disconnected before. But that is racy, until we can reliably detect that it *is* indeed the same keyboard. Doing that is a much bigger undertaking, see my recent comment in #323.

File: src/api/flash/index.js [hunk 1]
```diff
@@ -27,14 +27,114 @@ import { getStaticPath } from "../../renderer/config";
...
+          await delay(timeouts.dtrToggle);
+          resolve();
+        });
+      });
+    };
+
+    // Attempt calling device.reset first.
+    // If the firmware supports it, we'll reboot quickly. If it does not, we'll
+    // fall back to the serial HUP below.
+    await focus.command("device.reset");
```
[the remark this answers is not in the document]
Comment (algernon) [reply src/api/flash/index.js:64]: > Instead can we use probe to see if reset is supported?

That was my first idea too, but nothing guarantees that it works. For example, `device.reset` on my freshly baked Model01-sketch doesn't do the reset. It's there, it is advertised, it doesn't work. That's a bug, obviously, but bugs happen, and if we don't do the 1200 HUP dance, I wouldn't be able to flash firmware with Chrysalis. Not even the fixed firmware.

We could put this behind an option, but that's more hassle than its worth, imo.

[the remark this answers is not in the document]
Comment (obra) [reply src/api/flash/index.js:64]: hrm. in general, I feel like we shouldn’t be executing commands that don’t exist on the keyboard.

[the remark this answers is not in the document]
Comment (algernon) [reply src/api/flash/index.js:64]: Oh, I understood you're asking if we should skip the 1200 HUP dance if device.reset is present! If you think we should skip `device.reset` if it's not available, we can absolutely do that.

Give me a few moments and I'll push an update.

**Pull request merged**